staging/src/k8s.io/mount-utils/mount_helper_common.go

```diff
@@ -53,7 +53,7 @@ func CleanupMountWithForce(mountPath string, mounter MounterForceUnmounter, exte
 	}
 	var notMnt bool
 	var err error
-	if !corruptedMnt {
+	if !mounter.CanSafelySkipMountPointCheck() && !corruptedMnt {
```

Comment:
I am curious; why do you not need to send the mountPath here?

Reply:
In my understanding, whether the `mounter.CanSafelySkipMountPointCheck()` returns true is decided by [`detectSafeNotMountedBehaviorWithExec`](https://github.com/kubernetes/kubernetes/pull/109676/files/7e674280561f4bbc34f48a3fda2281d2978fbf36#diff-5a53703b1baddb2caa650077f26075c59361575c84387e1e9ad5570a738dd473R240-R255). And `detectSafeNotMountedBehaviorWithExec` seems to decide if it can safely skip mount-point check by checking if umount command to temp directory (/tmp/kubelet-detect-safe-umount) shows "not mounted".

Then, the question is "Are umount commands used in a system always the same for all the paths?"  As "EXTERNAL HELPERS" section of [umount(8) man page](https://man7.org/linux/man-pages/man8/umount.8.html) describes, it seems that external helpers like `umount.fuse` may be called. Therefore, just checking temp directory may not always return the behaviors of the umount command for all the path in the system.

Reply:
After thinking about it a bit more, I started to think that the umount command that is used when nothing actually mounted on the path may be the same (If helper is decided by the type of the mounted filesystem, nothing mounted would mean no helper). "not mounted" message will only appears in such a situation, so we may be able to rely on this behavior and regard it as mountPath agnostic.

I'm not 100% sure about it, so we will need to check the logic behind how helper is decided and check if there is no corner case.

Reply:
I'll add the `-i` flag to the `umount` call in `detectSafeNotMountedBehavior`:
```
       -i, --internal-only
           Do not call the /sbin/umount.filesystem helper even if it
           exists. By default such a helper program is called if it
           exists.
```

Which should be safe for the detection part.

For systems with helpers present, if the detection returns `true`, and later a mount path is passed to `CleanupMountPoint` that triggers a helper which somehow does not conform to the expected behavior, it will result in an error. The path will not be removed. I'm not sure whether this issue will surface in a kubelet event or not, but it seems like an acceptable (non-destructive) failure mode for this edge case.

Reply:
My tests show that the filesystem is resolved even when the target is not a mount point, when `umount` is executed by the root user.

My system has a single `ext4` filesystem at `/`; I created a helper at `/sbin/umount.ext4`:
```sh
#!/bin/sh

echo "ext4 helper reporting for duty! target: $1"
```

I see the desired `not mounted` output as a **non-root user**:
```
% LIBMOUNT_DEBUG=all umount $(mktemp -d) 
30067: libmount:     INIT: library debug mask: 0xffff
30067: libmount:     INIT: library version: 2.30.2
30067: libmount:     INIT:     feature: selinux
30067: libmount:     INIT:     feature: btrfs
30067: libmount:     INIT:     feature: assert
30067: libmount:     INIT:     feature: debug
Available "LIBMOUNT_DEBUG=<name>[,...]|<mask>" debug masks:
   all      [0xffff] : info about all subsystems
   cache    [0x0004] : paths and tags cache
   cxt      [0x0200] : library context (handler)
   diff     [0x0400] : mountinfo changes tracking
   fs       [0x0040] : FS abstraction
   help     [0x0001] : this help
   locks    [0x0010] : mtab and utab locking
   loop     [0x2000] : loop devices routines
   options  [0x0008] : mount options parsing
   tab      [0x0020] : fstab, mtab, mountinfo routines
   update   [0x0080] : mtab, utab updates
   utils    [0x0100] : misc library utils
   monitor  [0x0800] : mount tables monitor
   btrfs    [0x1000] : btrfs specific routines
30067: libmount:      CXT: [0x55d2c4cf18e0]: ----> allocate [RESTRICTED]
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: /tmp/tmp.NoOgpNRbgU
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: lookup FS for '/tmp/tmp.NoOgpNRbgU'
30067: libmount:      CXT: [0x55d2c4cf18e0]: checking for writable tab files
30067: libmount:    UTILS: utab: /run/mount/utab
30067: libmount:    CACHE: [0x55d2c4cf2ad0]: alloc
30067: libmount:    CACHE: [0x55d2c4cf2ad0]: canonicalize path /tmp/tmp.NoOgpNRbgU
30067: libmount:    CACHE: [0x55d2c4cf2ad0]: add entry [ 1] (path): /tmp/tmp.NoOgpNRbgU: /tmp/tmp.NoOgpNRbgU
30067: libmount:      CXT: [0x55d2c4cf18e0]: tabfilter ENABLED!
30067: libmount:      TAB: [0x55d2c4cf4730]: alloc
30067: libmount:      TAB: [0x55d2c4cf4730]: set table parser filter
30067: libmount:      TAB: [0x55d2c4cf4730]: mtab parse: #1 read mountinfo
30067: libmount:      TAB: [0x55d2c4cf4730]: /proc/self/mountinfo: start parsing [entries=0, filter=yes]
30067: libmount:       FS: [0x55d2c4cf4790]: free [refcount=0]
30067: libmount:      TAB: [0x55d2c4cf4730]: /proc/self/mountinfo: stop parsing (0 entries)
30067: libmount:      TAB: [0x55d2c4cf4730]: parsing done [filename=/proc/self/mountinfo, rc=0]
30067: libmount:      CXT: [0x55d2c4cf18e0]: mtab requested [nents=0]
30067: libmount:      TAB: [0x55d2c4cf4730]: unset table parser filter
30067: libmount:      CXT: [0x55d2c4cf18e0]: tabfilter disabled
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: mtab empty
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: cannot find '/tmp/tmp.NoOgpNRbgU' in mtab
30067: libmount:      CXT: [0x55d2c4cf18e0]: merging mount flags
30067: libmount:      CXT: [0x55d2c4cf18e0]: final flags: VFS=00000000 user=00000000
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: evaluating permissions
30067: libmount:      CXT: [0x55d2c4cf18e0]: cannot find /tmp/tmp.NoOgpNRbgU in mtab and you are not root
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount is not allowed for you
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: preparing failed
30067: libmount:      CXT: [0x55d2c4cf18e0]: excode: rc=1 message="not mounted"
umount: /tmp/tmp.NoOgpNRbgU: not mounted.
30067: libmount:      CXT: [0x55d2c4cf18e0]: <---- reset [status=0] ---->
30067: libmount:       FS: [0x55d2c4cf29f0]: free [refcount=0]
30067: libmount:      TAB: [0x55d2c4cf4730]: reset
30067: libmount:      TAB: [0x55d2c4cf4730]: free [refcount=0]
30067: libmount:      CXT: [0x55d2c4cf18e0]: <---- reset [status=0] ---->
30067: libmount:    CACHE: [0x55d2c4cf2ad0]: free [refcount=0]
30067: libmount:      CXT: [0x55d2c4cf18e0]: <---- free
```

However, as a **root user**, the filesystem is determined to be `ext4` and the helper is called:
```
% sudo LIBMOUNT_DEBUG=all umount $(mktemp -d)
28544: libmount:     INIT: library debug mask: 0xffff
28544: libmount:     INIT: library version: 2.30.2
28544: libmount:     INIT:     feature: selinux
28544: libmount:     INIT:     feature: btrfs
28544: libmount:     INIT:     feature: assert
28544: libmount:     INIT:     feature: debug
Available "LIBMOUNT_DEBUG=<name>[,...]|<mask>" debug masks:
   all      [0xffff] : info about all subsystems
   cache    [0x0004] : paths and tags cache
   cxt      [0x0200] : library context (handler)
   diff     [0x0400] : mountinfo changes tracking
   fs       [0x0040] : FS abstraction
   help     [0x0001] : this help
   locks    [0x0010] : mtab and utab locking
   loop     [0x2000] : loop devices routines
   options  [0x0008] : mount options parsing
   tab      [0x0020] : fstab, mtab, mountinfo routines
   update   [0x0080] : mtab, utab updates
   utils    [0x0100] : misc library utils
   monitor  [0x0800] : mount tables monitor
   btrfs    [0x1000] : btrfs specific routines
28544: libmount:      CXT: [0x55abb03ea8e0]: ----> allocate 
28544: libmount:      CXT: [0x55abb03ea8e0]: umount: /tmp/tmp.TBIZ7scVSe
28544: libmount:      CXT: [0x55abb03ea8e0]: checking for writable tab files
28544: libmount:    UTILS: utab: /run/mount/utab
28544: libmount:      CXT: [0x55abb03ea8e0]: enabling flag 0002
28544: libmount:      CXT: [0x55abb03ea8e0]: umount: mountinfo unnecessary [type=ext4]
28544: libmount:      CXT: [0x55abb03ea8e0]: merging mount flags
28544: libmount:      CXT: [0x55abb03ea8e0]: final flags: VFS=00000000 user=00000000
28544: libmount:      CXT: [0x55abb03ea8e0]: /sbin/umount.ext4         ... found
28544: libmount:      CXT: [0x55abb03ea8e0]: prepare update
28544: libmount:      CXT: [0x55abb03ea8e0]: skip update: NOMTAB flag
28545: libmount:      CXT: [0x55abb03ea8e0]: argv[0] = "/sbin/umount.ext4"
28545: libmount:      CXT: [0x55abb03ea8e0]: argv[1] = "/tmp/tmp.TBIZ7scVSe"
28545: libmount:      CXT: [0x55abb03ea8e0]: argv[2] = "-n"
ext4 helper reporting for duty! target: /tmp/tmp.TBIZ7scVSe
28544: libmount:      CXT: [0x55abb03ea8e0]: /sbin/umount.ext4 executed [status=0]
28544: libmount:      CXT: [0x55abb03ea8e0]: don't update: NOMTAB flag
28544: libmount:      CXT: [0x55abb03ea8e0]: excode: rc=0 message=""
28544: libmount:      CXT: [0x55abb03ea8e0]: <---- reset [status=1] ---->
28544: libmount:       FS: [0x55abb03ea9e0]: free [refcount=0]
28544: libmount:      CXT: [0x55abb03ea8e0]: <---- reset [status=0] ---->
28544: libmount:      CXT: [0x55abb03ea8e0]: <---- free
```

I'm still trying to understand why `ext4` is decided as the filesystem type, this seems to happen in this call (or the next line): https://github.com/util-linux/util-linux/blob/master/libmount/src/context_umount.c#L281

Reply:
Yes. The umount. suffix will be called if there is one under /sbin/ . I don't know for ext4. But for nfs and nfs4, umount.nfs or umount.nfs4 will all check if the mount point is in the mtab and is nfs type file system (https://github.com/stefanha/nfs-utils/blob/vsock/utils/mount/nfsumount.c#L284)

Reply:
FYI, @yuefido is on the EFS team -- I asked him to take a look at this because he has additional context on how this PR impacts our (EKS) users.